### PR TITLE
Feat: add csv to Tree solution

### PR DESCRIPTION
**Pull request by rei-dumand**
### Proposed Solution
Hi there, thanks for your time today. I have since implemented the following function:
- **csvToTree**
   - takes in a csv filepath;
   - returns a tree structure (object);

Which uses the following helpers:
- **parseCSV**
   - takes in a csv filepath;
   - returns a promise of the parsed csv (array).
- **rowToTree**
   - takes in the parsed csv data (array);
   - returns a tree structure (object).
- **addRowToTree**
   - takes in a tree node and an array of the keys to insert;
   - successively inserts keys in the tree and recurses by depth+1 between key inserts;
   - returns <void>;
   - recursive function that calls itself until there are no more keys to insert;
   - helper function of **rowToTree**.

### Notes
- Some csv entries are duplicates (i.e Nick from rows 34 & 35). I removed duplicates by checking the indName in my **addRowToTree** function. Could checking for duplicates before building the tree be more efficient?
- Some 'Role Name' have combined titles (i.e Giles "CFO/Company Secretary" from row 26). Not separating them means that for instance, Giles is marked in the output tree as "CFO/Company Secretary" & "CFO". If they reflect two dinstinct work periods and should be kept that way, that fine. If "CFO" should be mentioned only once, then I would need to implement a function to separate combined titles at "/" and modify my **addRowToTree** & **rowToTree** helpers to make them accept more than one 'Role Name.

### Next Steps
- Going to spend this evening of tomorrow morning looking into d3.js to visualise this tree.

### Run
To run the proposed solution in Node.js:
1. Install the dependencies with `npm i`
2. Run `npm run test`

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    for (let row of data) {
+        const [indName, orgName, indRole, index, sector] = row;
+        addRowToTree(tree, [index, sector, orgName, indRole, indName])
+    }
+    return tree;
+};
+
+function addRowToTree(currentNode, remainingKeys) {
+    let key = remainingKeys.shift()
+    if (remainingKeys.length === 1) {
```
Comment (wayneplural) [inline index.js:26]: I think this might break if you started with 1 key (or 0)

File: index.js [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        if (!currentNode[key]) {
+            currentNode[key] = remainingKeys;
+            return;
+        }
+        for (let index in currentNode[key]) {
+            if (currentNode[key][index] == remainingKeys[0]) return;
```
Comment (wayneplural) [inline index.js:32]: Prefer triple equals unless you know you need double equals, it can result in unexpected behaviour!

File: index.js [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    let tree = {};
+    data.shift();
+    for (let row of data) {
+        const [indName, orgName, indRole, index, sector] = row;
+        addRowToTree(tree, [index, sector, orgName, indRole, indName])
+    }
+    return tree;
+};
+
+function addRowToTree(currentNode, remainingKeys) {
```
Comment (wayneplural) [inline index.js:24]: I think this can be simplified but need to have a closer look. Will have a look later

**Review by wayneplural — COMMENTED** [review]
Sorry it took so long to look at. I started doing it when you sent it through originally and got sidetracked! I have added a few initial comments, would like to do more but want to make sure I submit what I have done so may add more comments later.

**Comment by rei-dumand** [comment]
Thank you for your feedback @wayneplural.
- Moving forward I will favour const instead of let.
- Same goes for === over ==.
- While I did not manage to simplify the addRowToTree function, I tinkered around making the desired tree structure live outside of the RowsToTree function. Simple destructuring such as below has the problem of hardcoding the data headers. 
```
const [indName, orgName, indRole, index, sector] = row;
addRowToTree(tree, [index, sector, orgName, indRole, indName])
``` 
- I also attempted to weed out some edge cases, warning the use when invalid or incomplete csv data is provided. However, addRowToTree doesn't cater for situations where only one key is provided by the user. It doesn't break but the output tree doesn't look reasonable, as seen below:
```
// Given only one key is provided:
{ 'Entry1': [], Entry2: [], Entry3: [] }
```
- Finally, in the future I'll make sure I make a separate branch for PRs.

**Pull request closed without merging**